accepted/future-releases/spread-collections/feature-specification.md

```diff
@@ -331,18 +331,28 @@ An expression like:
 ```
 
 Is syntactically parsed as `mapOrSetLiteral`. To determine whether it actually
-is a map or set, the surrounding context is used. Given an `mapOrSetLiteral`
-with context type C:
+is a map or set, the upwards and downwards type inference contexts are used.
 
-*   If `Set<Null>` is assignable to C, and `Map<Null, Null>` is not assignable
-    to C, then the collection is a set literal.
+Given a `mapOrSetLiteral`:
 
-*  Otherwise, it is a map literal.
+1.  If `Map<Null, Null>` is assignable to the literal's context type, then it
+    is a map literal.
+
+2.  Else, if `Set<Null>` is assignable to the literal's context type, then the
+    collection is a set literal.
+
+3.  Else, if it has spread elements and all of the spread element expression
```

Comment:
I would phrase this differently.  I think it's probably better to say that that *if* there is a non-empty context, you resolve based on the context type (as above, or however we decide).   Only if there is no context type do you resolve based on the upwards inference.  Otherwise, I think it's guaranteed to be an error anyway, and you're just making things confusing by still trying to use upwards inference.

Reply:
Agree. You only need to address the case where there is no context type and no type arguments. The rest are already handled and is unaffected by spreads or comprehensions.

I think we need to rewrite everything around this so that we only have "setOrMapLiteral" as a syntactic category. It's going to be complicated to have:
```
setLiteral ::= `{` setElements `}`
mapLiteral ::= `{ mapEntries `}`
mapEntry ::= expression `:` expression |  `...` expression | `if` `(` expression `)` mapEntry (`else` mapEntry)? | `for` `(` ... `in` expression `)` mapEntry ;
setElement ::= expression  |  `...` expression | `if` `(` expression `)` setElement (`else` setElement)? | `for` `(` ... `in` expression `)` setElement ;
setOrMapPart ::= |  `...` expression | `if` `(` expression `)` setOrMapPart (`else` setOrMapPart)? | `for` `(` ... `in` expression `)` setOrMapPart ;
```
because they *do* overlap unless deep-down in the structure there is either an `e` element or an `e1:e2` entry.

So, I'd go for unifying sets and maps, and then use the context type or content parts to determine which one it is. And I wouldn't separate spreads and if/for constructs, they both need the same treatment anyway.

Let me try doing that.